### DISCUSSION TITLE
Need help with odd SHAP KeyError

### DISCUSSION THREAD

**Discussion by fraseralex96** [shap/shap#3178] [Q&A]
Hi,

Can someone please help explain why I am receiving the following error message:

`---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File /share/apps/centos7/anaconda3/2022.05/lib/python3.9/site-packages/pandas/core/indexes/base.py:3621, in Index.get_loc(self, key, method, tolerance)
   3620 try:
-> 3621     return self._engine.get_loc(casted_key)
   3622 except KeyError as err:

File /share/apps/centos7/anaconda3/2022.05/lib/python3.9/site-packages/pandas/_libs/index.pyx:136, in pandas._libs.index.IndexEngine.get_loc()

File /share/apps/centos7/anaconda3/2022.05/lib/python3.9/site-packages/pandas/_libs/index.pyx:163, in pandas._libs.index.IndexEngine.get_loc()

File pandas/_libs/hashtable_class_helper.pxi:5198, in pandas._libs.hashtable.PyObjectHashTable.get_item()

File pandas/_libs/hashtable_class_helper.pxi:5206, in pandas._libs.hashtable.PyObjectHashTable.get_item()

KeyError: 5802

The above exception was the direct cause of the following exception:

KeyError                                  Traceback (most recent call last)
Input In [75], in <cell line: 2>()
      1 # make a shap value array for the prediction
----> 2 shaps = explainer.shap_values(np.array(X_test.iloc[0]))

File ~/.local/lib/python3.9/site-packages/shap/explainers/_gradient.py:144, in Gradient.shap_values(self, X, nsamples, ranked_outputs, output_rank_order, rseed, return_variances)
    106 def shap_values(self, X, nsamples=200, ranked_outputs=None, output_rank_order="max", rseed=None, return_variances=False):
    107     """ Return the values for the model applied to X.
    108 
    109     Parameters
   (...)
    142         were chosen as "top".
    143     """
--> 144     return self.explainer.shap_values(X, nsamples, ranked_outputs, output_rank_order, rseed, return_variances)

File ~/.local/lib/python3.9/site-packages/shap/explainers/_gradient.py:297, in _TFGradient.shap_values(self, X, nsamples, ranked_outputs, output_rank_order, rseed, return_variances)
    295         else:
    296             x = X[l][j]
--> 297         samples_input[l][k] = t * x + (1 - t) * self.data[l][rind]
    298         samples_delta[l][k] = x - self.data[l][rind]
    300 # compute the gradients at all the sample points

File /share/apps/centos7/anaconda3/2022.05/lib/python3.9/site-packages/pandas/core/frame.py:3505, in DataFrame.__getitem__(self, key)
   3503 if self.columns.nlevels > 1:
   3504     return self._getitem_multilevel(key)
-> 3505 indexer = self.columns.get_loc(key)
   3506 if is_integer(indexer):
   3507     indexer = [indexer]

File /share/apps/centos7/anaconda3/2022.05/lib/python3.9/site-packages/pandas/core/indexes/base.py:3623, in Index.get_loc(self, key, method, tolerance)
   3621     return self._engine.get_loc(casted_key)
   3622 except KeyError as err:
-> 3623     raise KeyError(key) from err
   3624 except TypeError:
   3625     # If we have a listlike key, _check_indexing_error will raise
   3626     #  InvalidIndexError. Otherwise we fall through and re-raise
   3627     #  the TypeError.
   3628     self._check_indexing_error(key)

KeyError: 5802`

The code used to run SHAP is as follows:

`    
    explainer = shap.GradientExplainer(model5, X_train)

    shaps = explainer.shap_values(np.array(X_test.iloc[0]))
`

The architecture for model5 is as follows:

 `   
    # set layer weights initialiser
    initializer = keras.initializers.GlorotUniform(seed=seed)
    
    # drug-cell line data input
    x_input = layers.Input(shape=(X_train.shape[1]))

    # 5 fully connected layers with batch norm and dropout
    x = layers.Dense(512, kernel_initializer=initializer, activation='relu')(x_input)
    x = layers.BatchNormalization()(x)
    x = layers.Dropout(0.1)(x)
    x = layers.Dense(1024, kernel_initializer=initializer, activation='relu')(x)
    x = layers.BatchNormalization()(x)
    x = layers.Dropout(0.1)(x)
    x = layers.Dense(2048, kernel_initializer=initializer, activation='relu')(x)
    x = layers.BatchNormalization()(x)
    x = layers.Dropout(0.1)(x)
    x = layers.Dense(1024, kernel_initializer=initializer, activation='relu')(x)
    x = layers.BatchNormalization()(x)
    x = layers.Dropout(0.1)(x)
    x = layers.Dense(512, kernel_initializer=initializer, activation='relu')(x)
    x = layers.BatchNormalization()(x)
    x = layers.Dropout(0.1)(x)
    
    output = layers.Dense(1, kernel_initializer=initializer)(x) # actiation layer is 1 neuron for single regression prediction value
        
    model = keras.Model(x_input, output)
    
    model.compile(optimizer=keras.optimizers.RMSprop(learning_rate=learning_rate, 
                                                        momentum=momentum), 
                                                        loss='mse', metrics=['mae'])
    
    `

X_test is an array of omic data shaped (3027, 1378)

**Comment by znacer** [comment]
I feel like there is an error in this line : 
`shaps = explainer.shap_values(np.array(X_test.iloc[0]))`
Have you try with something like : `shaps = explainer.shap_values(np.array(X_test.iloc[:1]))` or `shaps = explainer.shap_values(np.array([X_test.iloc[0]]))` ?

**Reply by fraseralex96** [reply]
Hi,

Thank you for the suggestion.

Unfortunately it still seems to throw the same error message

**Reply by znacer** [reply]
I tried a minimal reproduction of your case :  
```py
import keras
from keras import layers
import numpy as np
import pandas as pd
import shap
shap.initjs()

X_train = pd.DataFrame(np.random.rand(10, 10))
y_train = pd.DataFrame(np.random.rand(10, 1))
X_test = pd.DataFrame(np.random.rand(10, 10))
initializer = keras.initializers.RandomNormal(mean=0.0, stddev=0.05, seed=None)
learning_rate = 0.001
momentum = 0.9

# drug-cell line data input
x_input = layers.Input(shape=(X_train.shape[1]))

# 5 fully connected layers with batch norm and dropout
x = layers.Dense(512, kernel_initializer=initializer, activation='relu')(x_input)
x = layers.BatchNormalization()(x)
x = layers.Dropout(0.1)(x)
x = layers.Dense(1024, kernel_initializer=initializer, activation='relu')(x)
x = layers.BatchNormalization()(x)
x = layers.Dropout(0.1)(x)
x = layers.Dense(2048, kernel_initializer=initializer, activation='relu')(x)
x = layers.BatchNormalization()(x)
x = layers.Dropout(0.1)(x)
x = layers.Dense(1024, kernel_initializer=initializer, activation='relu')(x)
x = layers.BatchNormalization()(x)
x = layers.Dropout(0.1)(x)
x = layers.Dense(512, kernel_initializer=initializer, activation='relu')(x)
x = layers.BatchNormalization()(x)
x = layers.Dropout(0.1)(x)

output = layers.Dense(1, kernel_initializer=initializer)(x) # actiation layer is 1 neuron for single regression prediction value
    
model = keras.Model(x_input, output)

model.compile(optimizer=keras.optimizers.RMSprop(learning_rate=learning_rate, 
                                                    momentum=momentum), 
                                                    loss='mse', metrics=['mae'])
model.fit(X_train, y_train, epochs=1, batch_size=32, verbose=0)



# explain the model's predictions using SHAP values
explainer = shap.GradientExplainer(model, X_train)
shaps = explainer.shap_values(np.array([X_test.iloc[0]]))
```

I have no error so far. Maybe check that the number of columns of X_test and X_train are the same.

**Reply by fraseralex96** [reply]
Hi,

Yes I think you are right and it is something specific to the dataframes as I re-ran everything with randomly assigned values for the dataframes as you have above and it appears to work.

Both X_test and X_train have the same number of columns, and I have used these dataframes to run SHAP successfully with different models, it is bizarre that they don't work now.
